pycloudlib/azure/cloud.py

```diff
@@ -161,6 +170,26 @@ def __init__(
 
         self.resource_group = self._create_resource_group()
         self.base_tag = tag
+        self._enable_boot_diagnostics = enable_boot_diagnostics
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Get azure diagnostics and then cleanup."""
+        if exc_type or self._enable_boot_diagnostics:
+            for instance in self.created_instances:
+                boot_diagnostics = self.get_boot_diagnostics(instance)
+                # min. delay for the url to be active.
+                time.sleep(60)
+                if boot_diagnostics:
+                    response = requests.get(boot_diagnostics.console_log_url)
+                    filename = f"{instance.name}.boot_diagnostics"
+                    with open(filename, "w", encoding="utf-8") as f:
```

Comment:
As discussed in stand up, I do not agree with writing the boot diagnostics to file by default. It diverges from how the library has been handling logs and runtime information (the only other use of `open` is within `_download_file` and for handling ssh keys).

I'm in the camp of libraries should not directly write logging artifacts. Simply download the contents of the console URIs with one of two options:

1. Dump directly to logging and defer to the application logging handlers to access the logs.
2. Store the logs as part of the `Azure` class, and implement the appropriate exception handling for `_create_virtual_machine` to store the logs as part of the Azure cloud and allow the application to use the logs as it sees fit.

Ideally you would do both. Applications can either be lazy and use a handler to access the logs, or have full introspection into the logs at runtime and decide what to do with them then (with the benefit of not having a library arbitrarily writing files).